### PR TITLE
[SPARK-35162][SQL] New SQL functions: TRY_ADD/TRY_DIVIDE

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Add New SQL functions: 
* TRY_ADD
* TRY_DIVIDE

These expressions are identical to the following expression under ANSI mode except that it returns null if error occurs:
* ADD
* DIVIDE

Note: it is easy to add other expressions like `TRY_SUBTRACT`/`TRY_MULTIPLY` but let's control the number of these new expressions and just add `TRY_ADD` and `TRY_DIVIDE` for now.

### Why are the changes needed?

1. Users can manage to finish queries without interruptions in ANSI mode.
2. Users can get NULLs instead of unreasonable results if overflow occurs when ANSI mode is off.
For example, the behavior of the following SQL operations is unreasonable:
```
2147483647 + 2 => -2147483647
```

With the new safe version SQL functions:
```
TRY_ADD(2147483647, 2) => null
```

Note: **We should only add new expressions to important operators, instead of adding new safe expressions for all the expressions that can throw errors.** 
### Does this PR introduce _any_ user-facing change?

Yes, new SQL functions: TRY_ADD/TRY_DIVIDE

### How was this patch tested?

Unit test

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Try.scala [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+  override def dataType: DataType = child.dataType
+
+  override def nullable: Boolean = true
+
+  override protected def withNewChildInternal(newChild: Expression): Expression =
+    copy(child = newChild)
+
+  override def sql: String = s"try${child.sql}"
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Try.scala:55]: I will fix the `sql`/`toString` method. This is still WIP.

**Comment by gengliangwang** [comment]
@cloud-fan @maropu Sorry for the delay. This one is ready for review now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 2]
```diff
@@ -320,6 +320,8 @@ object FunctionRegistry {
     expression[Stack]("stack"),
     expression[CaseWhen]("when"),
 
+    expression[TryAdd]("try_add"),
+    expression[TryDivide]("try_divide"),
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:324]: If these functions exist in the `math` function group, how about moving them after L376?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, CodeGenerator, ExprCode}
+import org.apache.spark.sql.catalyst.expressions.codegen.Block._
+import org.apache.spark.sql.types.DataType
+
+private[catalyst] case class TryEval(child: Expression)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala:25]: we need `private[catalyst]` here? I think `catalyst` package itself looks private.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  override protected def withNewChildInternal(newChild: Expression): Expression =
+    copy(child = newChild)
+}
+
+@ExpressionDescription(
+  usage = "_FUNC_(expr1, expr2) - Returns `expr1`+`expr2` and the result is null on overflow.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(1, 2);
+       3
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala:62]: How about adding an overflow example here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(expr1, expr2) - Returns `expr1`/`expr2`. It always performs floating point division. Its result is always null if `expr2` is 0.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(3, 2);
+       1.5
+      > SELECT _FUNC_(2L, 2L);
+       1.0
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala:89]: How about adding a null example here?

**Review by maropu — COMMENTED** [review]
Just out of curiosity; Any reason to pick up try_add+try_divide instead of try_add+try_multiple?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 2 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala (hunk 3 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala (hunk 4 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala:61]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala (hunk 5 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala:90]: Done

**Comment by gengliangwang** [comment]
> Just out of curiosity; Any reason to pick up try_add+try_divide instead of try_add+try_multiple?

IMO, divide by 0 error is more common in ETL/ML jobs than integral multiply overflow. I also talked to @bart-samwel, from his experience on Bigquery, the most desired functions are `try_cast` and `try_divide`.
We can add `TRY_SUBTRACT`/`TRY_MULTIPLY` if many users want it. Before that, let's be cautious in adding such new functions.

**Review by maropu — APPROVED** [review]
Thank you for the explanation, @gengliangwang . Looks fine to me.

**Comment by gengliangwang** [comment]
Merging to master. @maropu Thanks for the review.

**Pull request closed without merging**